probe-rs/src/flashing/loader.rs

```diff
@@ -21,142 +21,195 @@ use crate::memory::MemoryInterface;
 use crate::session::Session;
 use crate::Target;
 
-/// `FlashLoader` is a struct which manages the flashing of any chunks of data onto any sections of flash.
-///
-/// Use [add_data()](FlashLoader::add_data) to add a chunk of data.
-/// Once you are done adding all your data, use `commit()` to flash the data.
-/// The flash loader will make sure to select the appropriate flash region for the right data chunks.
-/// Region crossing data chunks are allowed as long as the regions are contiguous.
-pub struct FlashLoader {
-    memory_map: Vec<MemoryRegion>,
-    builder: FlashBuilder,
-
-    /// Source of the flash description,
-    /// used for diagnostics.
-    source: TargetDescriptionSource,
+/// Helper trait for object safety.
+pub trait ImageReader: Read + Seek {}
+impl<T> ImageReader for T where T: Read + Seek {}
+
+/// Load and parse a firmware in a particular format, and add it to the flash loader.
+pub trait ImageLoader {
+    /// Loads the given image.
+    fn load(
+        &self,
+        flash_loader: &mut FlashLoader,
+        session: &mut Session,
+        file: &mut dyn ImageReader,
+    ) -> Result<(), FileDownloadError>;
 }
 
-impl FlashLoader {
-    /// Create a new flash loader.
-    pub fn new(memory_map: Vec<MemoryRegion>, source: TargetDescriptionSource) -> Self {
-        Self {
-            memory_map,
-            builder: FlashBuilder::new(),
-            source,
+impl ImageLoader for Format {
+    fn load(
+        &self,
+        flash_loader: &mut FlashLoader,
+        session: &mut Session,
+        file: &mut dyn ImageReader,
+    ) -> Result<(), FileDownloadError> {
+        match self {
+            Format::Bin(options) => BinLoader(options.clone()).load(flash_loader, session, file),
+            Format::Elf => ElfLoader.load(flash_loader, session, file),
+            Format::Hex => HexLoader.load(flash_loader, session, file),
+            Format::Idf(options) => IdfLoader(options.clone()).load(flash_loader, session, file),
```

Comment:
This is a different format for me, all the others just load binary data, but this actually does a transform. Is there a way to split this up maybe?

(At least I have already been confused by this, because I thought for a long time that this was a different input format...)

Reply:
Technically `bin` also does a transform

Reply:
So, technically, we are mixing image formats and platforms here. Probe-rs has originally worked with an implicit "bare-metal" platform assumption, and esp-idf came on top of that. This platform has a bootloader and a partition table, that we, in the image loading phase merge into the firmware, mainly for the convenience and to be beginner friendly. I guess we can, instead of hacking this into image formats, introduce a more explicit platform concept. The technical problems remain the same:
- registering a platform
- platform/MCU compatibility (esp-idf still won't make sense for an STM, even if some platforms may be totally vendor-agnostic [like an OS])
- platform-specific CLI/debugger options - needs dynamic clap and serde support to parse things like path to the bootloader file

While we could do all the packaging externally, it's really, really convenient for probe-rs to do automatically and really, really annoying to do externally, especially for beginners.

Reply:
I believe there is nothing inherent stopping us from flashing an esp-idf system from a bin or uf2, so separating these concepts might make sense. Either way I still don't like the tight coupled nature of image formats and I still think this extraction makes sense. But we can move into a direction more to your liking.

Reply:
For me, there are two things:

- The format of the file on disk, which is one of `hex`, `elf`, `bin` or `uf2`.
- A transformation that probe-rs applies to this, which is `esp-idf` for certain chips currently. I don't think changing the address in a binary file should really be counted here.

It would be nice if the `esp-idf` part would be completely indepent of an elf file, then this could be split up fairly cleanly I think.

Reply:
I appreciate your input, I think I agree with your views here and I will work towards something like this.